### PR TITLE
[SPARK-43386][SQL] Improve list of suggested column/attributes in `UNRESOLVED_COLUMN.WITH_SUGGESTION` error class

### PR DESCRIPTION
**Pull request by vitaliili-db**
### What changes were proposed in this pull request?

In this change we determine whether unresolved identifier is multipart or not and remap list of suggested columns to fit the same pattern.

- Main change is in `StringUtils.scala`. The rest is caused by method rename and test fixes.

### Why are the changes needed?

When determining a list of suggested columns/attributes for `UNRESOLVED_COLUMN.WITH_SUGGESTION` we sort by Levenshtein distance between unresolved identifier and list of fully qualified column names from target table. In case of a single-part identifier this might lead to poor experience especially for short(ish) identifiers, e.g. `a` and table with columns `m` and `aa` in default spark catalog => suggested list will be (`spark_catalog.default.m`, `spark_catalog.default.aa`) 

### Does this PR introduce _any_ user-facing change?

No, since we don't document internals of how we generate suggestion list for this error.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by srowen** [comment]
What is this?

**Comment by vitaliili-db** [comment]
@srowen fixed description, didn't expect you to be so quick

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala [hunk 1]
```diff
@@ -79,10 +80,38 @@ object StringUtils extends Logging {
...
       testStrings: Seq[String]): Seq[String] = {
-    testStrings.sortBy(LevenshteinDistance.getDefaultInstance.apply(_, baseString))
+    // This method is used to generate suggested list of candidates closest to `baseString` from the
+    // list of `testStrings`. Spark uses it to clarify error message in case a query refers to non
+    // existent column or attribute. The `baseString` could be single part or multi part and this
+    // method will try to match suggestions.
+    // Note that identifiers from `testStrings` could represent columns or attributes from different
+    // catalogs, schemas or tables. We preserve suggested identifier prefix and reconstruct
+    // multi-part identifier after ordering if there are more than one unique prefix in a list. This
+    // will also reconstruct multi-part identifier for the cases of nested columnns. E.g. for a
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:93]:
```suggestion
    // will also reconstruct multi-part identifier for the cases of nested columns. E.g. for a
```

Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:93]:
```suggestion
    // will also reconstruct multi-part identifier for the cases of nested columns. E.g. for a
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala [hunk 2]
```diff
@@ -79,10 +80,38 @@ object StringUtils extends Logging {
...
+    // table `t` with columns `a`, `b`, `c.d` (nested) and requested column `d` we will create
+    // prefixes `t`, `t`, and `t.c`. Since there is more than one distinct prefix we will return
+    // sorted suggestions as multi-part identifiers => (`t`.`c`.`d`, `t`.`a`, `t`.`b`).
+    val multiPart = UnresolvedAttribute.parseAttributeName(baseString).size > 1
+    if (multiPart) {
+      testStrings.sortBy(LevenshteinDistance.getDefaultInstance.apply(_, baseString))
+    } else {
+      val split = testStrings.map { ident =>
+        val parts = UnresolvedAttribute.parseAttributeName(ident)
+        (parts.take(parts.size - 1).map(quoteIfNeeded).mkString("."), quoteIfNeeded(parts.last))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:103]:
```suggestion
        val parts = UnresolvedAttribute.parseAttributeName(ident).map(quoteIfNeeded)
        (parts.init.mkString("."), parts.last)
```

**Review by MaxGekk — CHANGES_REQUESTED** [review]

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @vitaliili-db.

**Pull request closed without merging**